internal/datastore/proxy/relationshipintegrity.go

```diff
@@ -0,0 +1,397 @@
+	"github.com/authzed/spicedb/pkg/tuple"
+)
+
+// KeyConfig is a configuration for a key used to sign relationships.
+type KeyConfig struct {
+	ID        string
+	ExpiredAt *time.Time
+	Bytes     []byte
+}
+
+type hmacConfig struct {
+	keyID     string
+	expiredAt *time.Time
+	hmacPool  *sync.Pool
+}
+
+var (
+	alg         = sha256.New
+	versionByte = byte(0x01)
+)
+
+// NewRelationshipIntegrityProxy creates a new datastore proxy that ensures the integrity of
+// relationships by using HMACs to sign the data. The current key is used to sign new data,
+// and the expired keys are used to verify old data, if any.
+func NewRelationshipIntegrityProxy(ds datastore.Datastore, currentKey KeyConfig, expiredKeys []KeyConfig) (datastore.Datastore, error) {
+	currentKeyHMAC := hmacConfig{
+		keyID:     currentKey.ID,
+		expiredAt: currentKey.ExpiredAt,
+		hmacPool: &sync.Pool{
+			New: func() any { return hmac.New(alg, currentKey.Bytes) },
+		},
+	}
+
+	if currentKey.ExpiredAt != nil {
+		return nil, spiceerrors.MustBugf("current key cannot have an expiration")
+	}
+
+	keysByID := make(map[string]hmacConfig, len(expiredKeys)+1)
+	keysByID[currentKey.ID] = currentKeyHMAC
+
+	for _, key := range expiredKeys {
+		if key.ExpiredAt == nil {
+			return nil, spiceerrors.MustBugf("expired key missing expiration time")
+		}
+
+		if _, ok := keysByID[key.ID]; ok {
+			return nil, spiceerrors.MustBugf("found duplicate key ID: %s", key.ID)
+		}
+
+		keysByID[key.ID] = hmacConfig{
+			keyID:     key.ID,
+			expiredAt: key.ExpiredAt,
+			hmacPool: &sync.Pool{
+				New: func() any { return hmac.New(alg, key.Bytes) },
+			},
+		}
+	}
+
+	return &relationshipIntegrityProxy{
+		ds:         ds,
+		primaryKey: currentKeyHMAC,
+		keysByID:   keysByID,
+	}, nil
+}
+
+type relationshipIntegrityProxy struct {
+	ds         datastore.Datastore
+	primaryKey hmacConfig
+	keysByID   map[string]hmacConfig
+}
+
+func (r *relationshipIntegrityProxy) lookupKey(keyID string) (hmacConfig, error) {
+	key, ok := r.keysByID[keyID]
+	if !ok {
+		return hmacConfig{}, fmt.Errorf("key not found: %s", keyID)
+	}
+
+	return key, nil
+}
+
+// computeRelationshipHash computes the HMAC hash of a relationship tuple.
+func computeRelationshipHash(tpl *corev1.RelationTuple, key hmacConfig) ([]byte, error) {
+	hmac := key.hmacPool.Get().(hash.Hash)
+	defer key.hmacPool.Put(hmac)
+	defer hmac.Reset()
+
+	var sb strings.Builder
+	sb.WriteString(tpl.ResourceAndRelation.Namespace)
+	sb.WriteString(":")
+	sb.WriteString(tpl.ResourceAndRelation.ObjectId)
+	sb.WriteString("#")
+	sb.WriteString(tpl.ResourceAndRelation.Relation)
+	sb.WriteString("@")
+	sb.WriteString(tpl.Subject.Namespace)
+	sb.WriteString(":")
+	sb.WriteString(tpl.Subject.ObjectId)
+	sb.WriteString("#")
+	sb.WriteString(tpl.Subject.Relation)
+
+	if tpl.Caveat != nil && tpl.Caveat.CaveatName != "" {
+		sb.WriteString(" with ")
+		sb.WriteString(tpl.Caveat.CaveatName)
+		sb.WriteString(":")
+		sb.WriteString(tpl.Caveat.Context.String())
+	}
```

Comment:
is this something we can put in the tuple package? maybe something similar already exists

Reply:
Yes, we could put it into there